File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import qualified Data.Char as C
import qualified Data.Text as T

import Data.Text (Text)

responseFor :: Text -> Text
responseFor text
  | isSilence          = T.pack "Fine. Be that way!"
  | isShoutingQuestion = T.pack "Calm down, I know what I'm doing!"
  | isShouting         = T.pack "Whoa, chill out!"
  | isQuestion         = T.pack "Sure."
  | otherwise          = T.pack "Whatever."
  where
    isShouting         = T.any C.isAlpha text && T.toUpper text == text
    isQuestion         = T.isSuffixOf (T.pack "?") (T.stripEnd text)
    isShoutingQuestion = isShouting && isQuestion
    isSilence          = T.null (T.strip text)

File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(def silence? str/blank?)

(defn- upper?
  [text]
  (= (str/upper-case text) text))

(defn- question?
  [text]
  (str/ends-with? text "?"))

(defn- has-letter?
  [text]
  (some #(Character/isLetter %) text))

(defn- shouting?
  [text]
  (and (upper? text) (has-letter? text)))

(defn- shouting-question?
  [text]
  (and (shouting? text) (question? text)))

(defn response-for
  [input]
  (let [text (str/trimr input)]
    (cond
      (shouting-question? text) "Calm down, I know what I'm doing!"
      (silence? text) "Fine. Be that way!"
      (shouting? text) "Whoa, chill out!"
      (question? text) "Sure."
      :else "Whatever.")))